Added flag to ignore fluentd connect error on container start

jnummelin opened the pull request:
Added flag to ignore fluentd connect error during container start. The default is to fail the container start as with current log-drivers do.

Added also a flag to set the buffer limit for fluent client as that controls how much data the client will buffer while trying to reconnect.

Fixes #15757
Implements the flag as log-opt as requested in PR #17182

Signed-of-by: Jussi Nummelin jussi.nummelin@gmail.com

jnummelin committed:
Added flag to ignore fluentd connect error on container start

Signed-off-by: Jussi Nummelin <jussi.nummelin@gmail.com>

Changed buffer size to 1M and removed unnecessary fmt call

Signed-off-by: Jussi Nummelin <jussi.nummelin@gmail.com>

Updated docs for the new fluentd opts

Signed-off-by: Jussi Nummelin <jussi.nummelin@gmail.com>

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "feature/fluent-logger-ignore-connect-error-flag" git@github.com:jnummelin/docker.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

MHBauer commented on daemon/logger/fluentd/fluentd.go:
```diff
@@ -52,13 +54,24 @@ func New(ctx logger.Context) (logger.Logger, error) {
+	}
 	extra := ctx.ExtraAttributes(nil)
-	logrus.Debugf("logging driver fluentd configured for container:%s, host:%s, port:%d, tag:%s, extra:%v.", ctx.ContainerID, host, port, tag, extra)
+	logrus.Debugf("logging driver fluentd configured for container:%s, host:%s, port:%d, tag:%s, ignoreStartConnectError: %b,  extra:%v.", ctx.ContainerID, host, port, tag, ignoreStartConnectError, extra)
```
Do you want to print out buffer-limit as well here?

MHBauer commented on docs/reference/logging/fluentd.md:
```diff
@@ -34,6 +34,7 @@ Some options are supported by specifying `--log-opt` as many times as needed:
 
  - `fluentd-address`: specify `host:port` to connect `localhost:24224`
  - `tag`: specify tag for fluentd message, which interpret some markup, ex `{{.ID}}`, `{{.FullID}}` or `{{.Name}}` `docker.{{.ID}}`
+ - `ignore-startup-connect-error`: true/false; Should the logging driver ignore the error on container startup and reconnect in the background
```
Add `buffer-limit` as well?

calavera commented:
**This PR is signed. Please, DO NOT review until everything is properly signed.**

/cc @MHBauer

jnummelin commented:
@calavera That's weird, I did sign it and even the commit details show sign-off for both commits. :/ Now the sign-off seems to be working. I anyway pushed update for the comments. :)

calavera commented:
LGTM.

@LK4D4 can you take a look?

LK4D4 commented on daemon/logger/fluentd/fluentd.go:
```diff
@@ -132,3 +147,25 @@ func parseAddress(address string) (string, int, error) {
 	return host, portnum, nil
 }
+
+func parseStartConnectErrorFlag(ignoreConnectError string) (bool, error) {
```
this should be in `loggerutils` I think.

LK4D4 commented on daemon/logger/fluentd/fluentd.go:
```diff
@@ -52,13 +54,24 @@ func New(ctx logger.Context) (logger.Logger, error) {
+		if !ignoreStartConnectError {
+			return nil, err
+		}
+		logrus.Debugf("fluentd cannot connect to configured endpoint. Ignoring as instructed.")
```
I think it makes sense to log error here and on Warn level.

calavera replied:
I agree.

tagomoris replied:
:+1:

LK4D4 commented:
ping @tagomoris

tagomoris commented:
LGTM except for points already commented.

jnummelin commented:
I pushed update for the comments yesterday. Let me know if there's anything else that should be taken care of before merging.

thaJeztah commented:
@jnummelin it looks like this PR needs a rebase :cry:

thaJeztah commented:
Also, can you squash your commits, so that there's only one commit in this PR? You can find some instructions in our contributors guide; https://docs.docker.com/opensource/workflow/work-issue/#pull-and-rebase-frequently

jnummelin commented:
@thaJeztah Done and done :)

runcom commented:
not totatlly sold about `ignore-startup-connect-error` naming, LGTM otherwise

runcom commented:
ping @LK4D4

jnummelin commented:
@runcom @LK4D4 Any advice on the flag naming?

tiborvass commented:
@jnummelin do we really need a flag for this? Can't we have a built-in retry logic by default?

jnummelin commented:
@tiborvass On the other PR (already closed without merging) this was discussed and the consensus seemed to be to use a log-opt flag: https://github.com/docker/docker/pull/17182#issuecomment-156542543
For our use case the flag makes more sense as we can rely on the local buffering of the fluent driver. The retry logic would still imply that the container cannot start unless the log backend is available. In our case operating containers are more important than full logs but naturally that can change between use cases. Hence the `log-opt` flag so that people can change the behaviour case-by-case.

tiborvass commented on docs/reference/logging/fluentd.md:
```diff
@@ -34,7 +34,8 @@ Some options are supported by specifying `--log-opt` as many times as needed:
  - `fluentd-address`: specify `host:port` to connect `localhost:24224`
  - `tag`: specify tag for fluentd message, which interpret some markup, ex `{{.ID}}`, `{{.FullID}}` or `{{.Name}}` `docker.{{.ID}}`
-
+ - `ignore-startup-connect-error`: true/false; Should the logging driver ignore the error on container startup and reconnect in the background. Default: false
```
@jnummelin @runcom `reconnect-on-startup-error` ?

thaJeztah replied:
perhaps a similar approach as `--restart`? E.g. "always:10"? Not sure if more options are wanted

jnummelin replied:
@tiborvass `reconnect-on-startup-error` sounds good to me. Shall we lock that down as the final answer? :)

tiborvass replied:
@LK4D4 @calavera what do you think?

LK4D4 replied:
And true by default? Actually, I don't think that it really reflects flag nature. The main idea is not to connect or reconnect, but container start failure.

jnummelin replied:
What about `fail-on-startup-error`, default=´true´? That reflects the flag nature pretty well. We could get even more descriptive like `fail-container-start-on-connect-error` but then the flag becomes just too long to be usable. :)

LK4D4 replied:
`--fcsoce` :D Yeah, I'm ok with `fail-on-startup-error`

jnummelin replied:
Ok, I'll change the PR asap

jnummelin commented:
Updated the flag as discussed in previous comments. The build errors do not seem to have anything to do with these changes. The branch is also rebased.

jnummelin commented:
Any advice how to push this forward?

cpuguy83 commented on daemon/logger/fluentd/fluentd.go:
```diff
@@ -24,10 +24,12 @@ type fluentd struct {
+	defaultPort                      = 24224
+	defaultTagPrefix                 = "docker"
+	defaultIgnoreConnectErrorOnStart = false           // So that we do not break existing behaviour
+	defaultBufferLimit               = 8 * 1024 * 1024 // 8M buffer by default
```
A default 1MB buffer seems a bit more sane here.

jnummelin replied:
The fluent driver itself uses 8M, that's why I chose it also.
https://github.com/fluent/fluent-logger-golang/blob/master/fluent/fluent.go#L21

For me it seems appropriate to use the same default as the fluent driver but I can change it to 1M if that is still seen as better option?

cpuguy83 replied:
It doesn't seem like much, until you have 100+ containers all buffering 8MB due to an outage.
I think we should change it for our use-case here.

jnummelin replied:
Makes sense, I'll update the PR soon.

cpuguy83 commented on daemon/logger/loggerutils/log_option_helpers.go:
```diff
@@ -0,0 +1,27 @@
+// the driver should ignore possible connection errors during startup
+func ParseFailOnStartupErrorFlag(ctx logger.Context) (bool, error) {
+	failOnStartupError := ctx.Config["fail-on-startup-error"]
+	fmt.Printf("%s", failOnStartupError)
```
Remove `fmt` call

jnummelin commented:
Changed the code as @cpuguy83 commented. Anything else still needed?

cpuguy83 commented:
LGTM

calavera commented:
LGTM

vdemeester commented on docs/reference/logging/fluentd.md:
```diff
@@ -34,7 +34,8 @@ Some options are supported by specifying `--log-opt` as many times as needed:
  - `fluentd-address`: specify `host:port` to connect `localhost:24224`
  - `tag`: specify tag for fluentd message, which interpret some markup, ex `{{.ID}}`, `{{.FullID}}` or `{{.Name}}` `docker.{{.ID}}`
-
+ - `fail-on-startup-error`: true/false; Should the logging driver fail container startup in case of connect error during startup. Default: true (backwards compatible)
```
These options should be detailed in `## Options` too (just below : https://github.com/docker/docker/pull/18041/files#diff-9a6be46aa59594b9640ede14e4bcc8acR59) and in more detail (or these a little more summarized — e.g. default value, etc.).

vdemeester commented:
@jnummelin the file `docs/reference/logging/overview.md` (https://github.com/docker/docker/blob/master/docs/reference/logging/overview.md#fluentd-options) should be update too.

jnummelin commented:
@vdemeester Updated the docs. PTAL

calavera commented:
LGTM.

Can we please rebase this and merge it? @vdemeester, @thaJeztah ?

jnummelin commented:
Rebased

thaJeztah commented:
@jnummelin can you squash your commits as well?

jnummelin commented:
@thaJeztah done

thaJeztah commented:
Thanks, LGTM

ping @vdemeester ptal

vdemeester commented:
LGTM 🐻

calavera commented:
clicked the wrong button by mistake :raised_hands: 

Merged :tada:
